src/storage/src/hummock/sstable/multi_builder.rs

```diff
@@ -146,10 +159,42 @@ where
         is_new_user_key: bool,
     ) -> HummockResult<()> {
         let mut switch_builder = false;
-        if self.split_by_table && full_key.user_key.table_id.table_id != self.last_table_id {
-            self.last_table_id = full_key.user_key.table_id.table_id;
-            switch_builder = true;
+        let key_table_id = full_key.user_key.table_id.table_id();
+        if key_table_id != self.last_table_id {
+            if self.split_by_table {
+                switch_builder = true;
+            }
+            while let Some((table_id, table_weight)) = self.table_weights.peek() && table_id < &&key_table_id {
+                if table_weight > &&0 {
+                    switch_builder = true;
+                }
+                self.table_weights.next();
+            }
+
+            if let Some((table_id, table_weight)) = self.table_weights.peek() && table_id == &&key_table_id && **table_weight > 1 {
+                self.vnode_split_point = VirtualNode::COUNT / (**table_weight as usize) - 1;
+            } else {
+                self.vnode_split_point = VirtualNode::MAX.to_index();
+            }
+        }
+        if self.vnode_split_point != VirtualNode::MAX.to_index() {
+            let key_vnode = full_key.user_key.get_vnode_id();
+            if key_vnode > self.vnode_split_point {
+                switch_builder = true;
+
+                let table_weight = *self.table_weights.peek().unwrap().1 as usize;
+                let (basic, remainder) = VirtualNode::COUNT.div_rem(&table_weight);
```

Comment:
Can table_weight be 0?

Reply:
> Can table_weight be 0?

When `table_weight==0`, it's unreachable.